File: frontend/src/app/core/services/conflicts/conflict.service.ts
import { Injectable, inject } from '@angular/core';
import { Employee } from '../employees/employee.types';
import { WorkHours } from '../schedule/schedule.types';
import { EmployeesService } from '../employees/employees.service';
import {ValidationResult} from './conflict.types';


@Injectable({
  providedIn: 'root'
})
export class ConflictService {
  private parseWorkHoursCache = new Map<string, {startTime: number, endTime: number}>();
  private dayNumberCache = new Map<string, number>();

  validateWorkHoursExceed12h(hoursString: string): ValidationResult | null {
    if (!hoursString) return null;

    const shift = this.parseWorkHours(hoursString);
    if (!shift) return null;

    const shiftLengthMinutes = shift.endTime - shift.startTime;
    const shiftLengthHours = shiftLengthMinutes / 60;

    if (shiftLengthHours > 12) {
      return {
        type: 'exceed12h',
        message: 'Uwzględnij odbior nadgodzin dla pracownika.'
      };
    }

    return null;
  }

  validateRestTimeConflicts(workHours: WorkHours[], employees: Employee[]): Set<string> {
    const conflicts = new Set<string>();
    const sortedWorkHours = [...workHours].sort((a, b) => a.date.localeCompare(b.date));
    const workHoursByEmployee = new Map<string, WorkHours[]>();

    sortedWorkHours.forEach(wh => {
      if (!workHoursByEmployee.has(wh.employee)) {
        workHoursByEmployee.set(wh.employee, []);
      }
      workHoursByEmployee.get(wh.employee)!.push(wh);
    });

    employees.forEach(employee => {
      const employeeWorkHours = workHoursByEmployee.get(employee.id) || [];

      for (let i = 0; i < employeeWorkHours.length - 1; i++) {
        const currentDay = employeeWorkHours[i];
        const nextDay = employeeWorkHours[i + 1];
        const restTime = this.calculateRestBetween(currentDay, nextDay);

        if (restTime < 11) {
          conflicts.add(`${employee.id}-${currentDay.date}`);
          conflicts.add(`${employee.id}-${nextDay.date}`);
        }
      }
    });

    return conflicts;
  }

  validate35HourRest(workHours: WorkHours[], employees: Employee[], currentMonthDate: Date): Map<string, Set<number>> {
    const daysInMonth = new Date(currentMonthDate.getFullYear(), currentMonthDate.getMonth() + 1, 0).getDate();
    const numberOfWeeks = Math.ceil(daysInMonth / 7);
    const sortedWorkHours = [...workHours].sort((a, b) => a.date.localeCompare(b.date));
    const workHoursByEmployee = new Map<string, WorkHours[]>();

    sortedWorkHours.forEach(wh => {
      if (!workHoursByEmployee.has(wh.employee)) {
        workHoursByEmployee.set(wh.employee, []);
      }
      workHoursByEmployee.get(wh.employee)!.push(wh);
    });

    const badWeeksMap = new Map<string, Set<number>>();

    employees.forEach(employee => {
      const employeeShifts = workHoursByEmployee.get(employee.id) || [];
      const employeeBadWeeks = new Set<number>();

      const shiftsByWeek = new Map<number, WorkHours[]>();
      employeeShifts.forEach(shift => {
        const dayNumber = this.getDayNumber(shift.date);
        const weekNumber = Math.ceil(dayNumber / 7);

        if (!shiftsByWeek.has(weekNumber)) {
          shiftsByWeek.set(weekNumber, []);
        }
        shiftsByWeek.get(weekNumber)!.push(shift);
      });

      for (let week = 1; week <= numberOfWeeks; week++) {
        const weekShifts = shiftsByWeek.get(week) || [];
        const restPeriods = this.calculateAllRestPeriodsInWeek(weekShifts, week, currentMonthDate);
        const hasLongRest = restPeriods.some(rest => rest.hours >= 35);

        if (!hasLongRest) {
          employeeBadWeeks.add(week);
        }
      }

      if (employeeBadWeeks.size > 0) {
        badWeeksMap.set(employee.id, employeeBadWeeks);
      }
    });

    return badWeeksMap;
  }

  private parseWorkHours(hoursString: string): { startTime: number, endTime: number } | null {
    const match = hoursString.match(/(\d{1,2}):(\d{2})-(\d{1,2}):(\d{2})/);
    if (!match) return null;

    const [, startHour, startMin, endHour, endMin] = match;
    return {
      startTime: parseInt(startHour) * 60 + parseInt(startMin),
      endTime: parseInt(endHour) * 60 + parseInt(endMin)
    };
  }

  private calculateRestBetween(currentDay: WorkHours, nextDay: WorkHours): number {
    const currentShift = this.parseWorkHoursCached(currentDay.hours);
    const nextShift = this.parseWorkHoursCached(nextDay.hours);

    if (!currentShift || !nextShift) return 24;

    const currentDate = new Date(currentDay.date);
    const nextDate = new Date(nextDay.date);

    const currentEndTime = new Date(currentDate);
    currentEndTime.setHours(
      Math.floor(currentShift.endTime / 60),
      currentShift.endTime % 60,
      0, 0
    );

    const nextStartTime = new Date(nextDate);
    nextStartTime.setHours(
      Math.floor(nextShift.startTime / 60),
      nextShift.startTime % 60,
      0, 0
    );

    const restHours = (nextStartTime.getTime() - currentEndTime.getTime()) / (1000 * 60 * 60);
    return Math.max(0, restHours);
  }

  private parseWorkHoursCached(hoursString: string): { startTime: number, endTime: number } | null {
    if (this.parseWorkHoursCache.has(hoursString)) {
      return this.parseWorkHoursCache.get(hoursString)!;
    }

    const parsed = this.parseWorkHours(hoursString);
    if (parsed) {
      this.parseWorkHoursCache.set(hoursString, parsed);
    }
    return parsed;
  }

  private getDayNumber(dateString: string): number {
    if (!this.dayNumberCache.has(dateString)) {
      this.dayNumberCache.set(dateString, new Date(dateString).getDate());
    }
    return this.dayNumberCache.get(dateString)!;
  }

  private calculateAllRestPeriodsInWeek(
    weekShifts: any[],
    weekNumber: number,
    currentMonthDate: Date
  ): Array<{hours: number}> {

    if (weekShifts.length === 0) {
      return [{hours: 168}];
    }

    const restPeriods: Array<{hours: number}> = [];
    const year = currentMonthDate.getFullYear();
    const month = currentMonthDate.getMonth();

    // Oblicz granice tygodnia
    const weekStart = (weekNumber - 1) * 7 + 1;
    const daysInMonth = new Date(year, month + 1, 0).getDate();
    const weekEnd = Math.min(weekNumber * 7, daysInMonth);


    // ✅ 1. Przerwa od początku tygodnia do pierwszej zmiany
    const firstShift = weekShifts[0];
    const firstShiftDay = new Date(firstShift.date).getDate();
    const firstShiftStart = this.parseShiftStartTime(firstShift, year, month);

    if (firstShiftDay > weekStart) {
      // Całe dni przed pierwszą zmianą
      const fullDaysBefore = firstShiftDay - weekStart;
      const hoursFromFullDays = fullDaysBefore * 24;
      restPeriods.push({hours: hoursFromFullDays});
    }

    if (firstShiftDay === weekStart && firstShiftStart) {
      // Godziny od początku dnia do pierwszej zmiany
      const startOfDay = new Date(year, month, firstShiftDay, 0, 0, 0);
      const hoursFromDayStart = (firstShiftStart.getTime() - startOfDay.getTime()) / (1000 * 60 * 60);
      if (hoursFromDayStart > 0) {
        restPeriods.push({hours: hoursFromDayStart});
      }
    }

    // ✅ 2. Przerwy między zmianami (istniejący kod)
    for (let i = 0; i < weekShifts.length - 1; i++) {
      const currentEnd = this.parseShiftEndTime(weekShifts[i], year, month);
      const nextStart = this.parseShiftStartTime(weekShifts[i + 1], year, month);

      if (currentEnd && nextStart) {
        const restHours = (nextStart.getTime() - currentEnd.getTime()) / (1000 * 60 * 60);

        if (restHours > 0) {
          restPeriods.push({hours: restHours});
        }
      }
    }

    // ✅ 3. Przerwa od ostatniej zmiany do końca tygodnia
    const lastShift = weekShifts[weekShifts.length - 1];
    const lastShiftDay = new Date(lastShift.date).getDate();
    const lastShiftEnd = this.parseShiftEndTime(lastShift, year, month);

    if (lastShiftDay < weekEnd) {
      // Całe dni po ostatniej zmianie
      const fullDaysAfter = weekEnd - lastShiftDay;
      const hoursFromFullDays = fullDaysAfter * 24;
      restPeriods.push({hours: hoursFromFullDays});
    }

    if (lastShiftDay === weekEnd && lastShiftEnd) {
      // Godziny od ostatniej zmiany do końca dnia
      const endOfDay = new Date(year, month, lastShiftDay, 23, 59, 59);
      const hoursToEndOfDay = (endOfDay.getTime() - lastShiftEnd.getTime()) / (1000 * 60 * 60);
      if (hoursToEndOfDay > 0) {
        restPeriods.push({hours: hoursToEndOfDay});
      }
    }

    return restPeriods;
  }

  private parseShiftStartTime(shift: any, year: number, month: number): Date | null {
    const match = shift.hours.match(/(\d{1,2}):(\d{2})-/);
    if (!match) return null;

    const date = new Date(year, month, new Date(shift.date).getDate());
    date.setHours(parseInt(match[1]), parseInt(match[2]), 0, 0);
    return date;
  }

  private parseShiftEndTime(shift: any, year: number, month: number): Date | null {
    const match = shift.hours.match(/-(\d{1,2}):(\d{2})/);
    if (!match) return null;

    const date = new Date(year, month, new Date(shift.date).getDate());
    date.setHours(parseInt(match[1]), parseInt(match[2]), 0, 0);
    return date;
  }

}

File: backend/backend/apps/schedule/services/conflict_service.py
"""
Serwis do wykrywania konfliktów w harmonogramach pracy.
"""
import re
from datetime import datetime, timedelta


from ..models import WorkHours
from ...employees.models import Employee


class ConflictDetectionService:
    """
    Serwis odpowiedzialny za wykrywanie konfliktów w harmonogramie pracy.
    """

    def __init__(self, location_id: str, month: int, year: int):
        """
        Inicjalizacja serwisu.

        Args:
            location_id: UUID lokacji
            month: Miesiąc (1-12)
            year: Rok
        """
        self.location_id = location_id
        self.month = month
        self.year = year
        self.work_hours = None
        self.employees = None

    def detect_all_conflicts(self):
        """
        Główna metoda - wykrywa wszystkie typy konfliktów.

        Returns:
            Dict z kluczami: rest_11h, rest_35h, exceed_12h
        """
        # Pobierz dane z bazy
        self._load_data()

        # Wykryj wszystkie konflikty
        conflicts = {
            'rest_11h': self._check_11h_rest(),
            'rest_35h': self._check_35h_rest(),
            'exceed_12h': self._check_12h_exceed()
        }

        return conflicts

    def _load_data(self):
        """Ładuje dane z bazy - work_hours i employees."""
        # Pobierz godziny pracy dla lokacji i miesiąca
        self.work_hours = WorkHours.objects.filter(
            location_id=self.location_id,
            date__month=self.month,
            date__year=self.year
        ).select_related('employee').order_by('date')

        # Pobierz pracowników na umowie o pracę dla tej lokacji
        employee_ids = self.work_hours.values_list('employee_id', flat=True).distinct()
        self.employees = Employee.objects.filter(
            id__in=employee_ids,
            agreement_type='permanent'  # Tylko UoP
        )

    def _check_11h_rest(self):
        """
        Sprawdza czy między zmianami jest przerwa 11h.

        Returns:
            Lista kluczy komórek z konfliktem (format: "employee_id-YYYY-MM-DD")
        """
        conflicts = set()

        # Pogrupuj godziny po pracownikach
        work_hours_by_employee = {}
        for wh in self.work_hours:
            if wh.employee_id not in work_hours_by_employee:
                work_hours_by_employee[wh.employee_id] = []
            work_hours_by_employee[wh.employee_id].append(wh)

        # Sprawdź każdego pracownika
        for employee_id, shifts in work_hours_by_employee.items():
            # Tylko dla pracowników na UoP
            if not self.employees.filter(id=employee_id).exists():
                continue

            # Posortuj zmiany po dacie
            shifts = sorted(shifts, key=lambda x: x.date)

            # Sprawdź każdą parę kolejnych zmian
            for i in range(len(shifts) - 1):
                current_shift = shifts[i]
                next_shift = shifts[i + 1]

                rest_hours = self._calculate_rest_between(current_shift, next_shift)

                if rest_hours < 11:
                    # Dodaj obie komórki do konfliktów
                    conflicts.add(f"{employee_id}-{current_shift.date}")
                    conflicts.add(f"{employee_id}-{next_shift.date}")

        return list(conflicts)

    def _check_35h_rest(self):
        """
        Sprawdza czy w każdym tygodniu jest przerwa 35h.
        UWAGA: Ostatni tydzień miesiąca jest pomijany (przerwa może być w następnym miesiącu).

        Returns:
            Dict: {employee_id: [week1, week2, ...]} - tygodnie z konfliktem
        """
        bad_weeks = {}

        # Oblicz liczbę tygodni w miesiącu
        days_in_month = self._get_days_in_month()
        num_weeks = (days_in_month + 6) // 7  # Zaokrąglenie w górę

        # Pogrupuj po pracownikach
        work_hours_by_employee = {}
        for wh in self.work_hours:
            if wh.employee_id not in work_hours_by_employee:
                work_hours_by_employee[wh.employee_id] = []
            work_hours_by_employee[wh.employee_id].append(wh)

        # Sprawdź każdego pracownika
        for employee_id, shifts in work_hours_by_employee.items():
            # Tylko dla pracowników na UoP
            if not self.employees.filter(id=employee_id).exists():
                continue

            employee_bad_weeks = []

            # Pogrupuj zmiany po tygodniach
            shifts_by_week = {}
            for shift in shifts:
                week_number = (shift.date.day + 6) // 7  # Tydzień 1-5
                if week_number not in shifts_by_week:
                    shifts_by_week[week_number] = []
                shifts_by_week[week_number].append(shift)

            # ✅ ZMIANA: Sprawdź każdy tydzień OPRÓCZ OSTATNIEGO
            for week in range(1, num_weeks):  # ← Zmienione z num_weeks + 1 na num_weeks
                week_shifts = shifts_by_week.get(week, [])

                if not week_shifts:
                    # Brak zmian = cały tydzień wolny (OK)
                    continue

                # Sprawdź czy jest przerwa 35h w tym tygodniu
                rest_periods = self._calculate_rest_periods_in_week(
                    week_shifts, week, days_in_month
                )

                has_long_rest = any(period >= 35 for period in rest_periods)

                if not has_long_rest:
                    employee_bad_weeks.append(week)

            if employee_bad_weeks:
                bad_weeks[str(employee_id)] = employee_bad_weeks

        return bad_weeks

    def _check_12h_exceed(self):
        """
        Sprawdza czy zmiana przekracza 12h.

        Returns:
            Lista kluczy komórek (format: "employee_id-YYYY-MM-DD")
        """
        exceeding = set()

        for wh in self.work_hours:
            # Tylko dla pracowników na UoP
            if not self.employees.filter(id=wh.employee_id).exists():
                continue

            shift_length = self._parse_shift_length(wh.hours)

            if shift_length and shift_length > 12:
                exceeding.add(f"{wh.employee_id}-{wh.date}")

        return list(exceeding)

    # === METODY POMOCNICZE ===

    def _parse_shift_length(self, hours_string: str):
        """
        Parsuje string godzin i zwraca długość zmiany w godzinach.

        Args:
            hours_string: Format "8:00-16:00"

        Returns:
            Długość zmiany w godzinach (float) lub None
        """
        match = re.match(r'(\d{1,2}):(\d{2})-(\d{1,2}):(\d{2})', hours_string)

        if not match:
            return None

        start_hour, start_min, end_hour, end_min = map(int, match.groups())

        start_minutes = start_hour * 60 + start_min
        end_minutes = end_hour * 60 + end_min

        shift_minutes = end_minutes - start_minutes
        return shift_minutes / 60

    def _calculate_rest_between(self, current_shift, next_shift):
        """
        Oblicza przerwę między dwiema zmianami w godzinach.
        """
        # Parsuj koniec obecnej zmiany
        current_match = re.match(r'.*-(\d{1,2}):(\d{2})', current_shift.hours)
        if not current_match:
            return 24  # Błąd parsowania = zakładamy OK

        current_end_hour, current_end_min = map(int, current_match.groups())

        # Parsuj początek następnej zmiany
        next_match = re.match(r'(\d{1,2}):(\d{2})-', next_shift.hours)
        if not next_match:
            return 24

        next_start_hour, next_start_min = map(int, next_match.groups())

        # Stwórz datetime
        current_end = datetime.combine(
            current_shift.date,
            datetime.min.time().replace(hour=current_end_hour, minute=current_end_min)
        )

        next_start = datetime.combine(
            next_shift.date,
            datetime.min.time().replace(hour=next_start_hour, minute=next_start_min)
        )

        # Oblicz różnicę
        rest_delta = next_start - current_end
        return rest_delta.total_seconds() / 3600  # Konwersja na godziny

    def _calculate_rest_periods_in_week(self, week_shifts, week_number: int, days_in_month: int):
        """
        Oblicza wszystkie okresy odpoczynku w tygodniu.

        Returns:
            Lista długości przerw w godzinach
        """
        if not week_shifts:
            return [168]  # Cały tydzień = 7 dni * 24h

        rest_periods = []

        # Posortuj zmiany
        week_shifts = sorted(week_shifts, key=lambda x: x.date)

        # Granice tygodnia
        week_start = (week_number - 1) * 7 + 1
        week_end = min(week_number * 7, days_in_month)

        # 1. Przerwa od początku tygodnia do pierwszej zmiany
        first_shift = week_shifts[0]
        if first_shift.date.day > week_start:
            # Całe dni przed pierwszą zmianą
            days_before = first_shift.date.day - week_start
            rest_periods.append(days_before * 24)

        # 2. Przerwy między zmianami
        for i in range(len(week_shifts) - 1):
            rest = self._calculate_rest_between(week_shifts[i], week_shifts[i + 1])
            rest_periods.append(rest)

        # 3. Przerwa od ostatniej zmiany do końca tygodnia
        last_shift = week_shifts[-1]
        if last_shift.date.day < week_end:
            days_after = week_end - last_shift.date.day
            rest_periods.append(days_after * 24)

        return rest_periods

    def _get_days_in_month(self):
        """Zwraca liczbę dni w miesiącu."""
        if self.month == 12:
            next_month = datetime(self.year + 1, 1, 1)
        else:
            next_month = datetime(self.year, self.month + 1, 1)

        last_day = next_month - timedelta(days=1)
        return last_day.day